### PR TITLE
Get pulse rate from file for slow mode

### LINKED ISSUE #35: Use frame times from file to dictate pulse rate for --slow mode
Requested by Torben

### PR DESCRIPTION
**Pull request by matthew-d-jones**
### Description of work

When in `--slow` mode, uses frame time dataset from `NXevent_data` to dictate how long the publisher waits between publishing the data for each pulse. This should work even if the time between pulses is not constant (for example at ISIS where some pulses are sent to the other target).

Now uses a struct for the optional args as many are passed to the publisher class.

### Issue

Closes #35

### Unit Tests

Added a unit test to `NexusFileReaderTest` for getting the relative frame time from file.

---

## Code Review (To be filled in by the reviewer only)

- [ ] Is the code of an acceptable quality?
- [ ] Do the changes function as described and is it robust?

---

## Nominate for Group Code Review (Anyone can nominate it)
Indicate if you think the code should be reviewed in a Thursday code review session.

- [ ] Recommend for group code review

Also, nominate it on the code_review Slack channel (does someone want to automate this?).

### CODE REVIEW

**Comment by matthew-d-jones** [comment]
@rerpha I'll add a unit test to cover --slow mode before asking you to take another look at this.

File: nexus_file_reader/src/NexusFileReader.cpp [hunk 1]
```diff
@@ -237,10 +236,26 @@ uint64_t NexusFileReader::getFrameTime(hsize_t frameNumber) {
...
 }
 
+/**
+ * Gets the frame time relative to the start of run, in milliseconds
+ *
+ * @param frameNumber - find the event index for the start of this frame
+ * @return - relative time of frame in milliseconds since run start
+ */
+uint64_t
+NexusFileReader::getRelativeFrameTimeMilliseconds(hsize_t frameNumber) {
```
Comment (rerpha) [inline nexus_file_reader/src/NexusFileReader.cpp]: parameter could be `const`

File: nexus_file_reader/test/HDF5FileTestHelpers.cpp [hunk 2]
```diff
@@ -10,6 +10,15 @@ hdf5::file::File createInMemoryTestFile(const std::string &filename) {
                             hdf5::property::FileCreationList(), fapl);
 }
 
+hdf5::file::File
+createInMemoryTestFileWithEventData(const std::string &filename) {
```
Comment (rerpha) [inline nexus_file_reader/test/HDF5FileTestHelpers.cpp]: parameter could be `const`

File: nexus_producer/test/NexusPublisherTest.cpp [hunk 3]
```diff
@@ -4,25 +4,63 @@
...
+    OptionalArgs settings;
+    settings.broker = "broker_name";
+    settings.instrumentName = "unitTest";
+    settings.quietMode = quiet;
+    settings.filename = testDataPath + "SANS_test_reduced.hdf5";
+    settings.detSpecFilename = testDataPath + "spectrum_gastubes_01.dat";
+    return settings;
+  }
+
+  NexusPublisher createStreamer(bool quiet) {
```
Comment (rerpha) [inline nexus_producer/test/NexusPublisherTest.cpp]: parameter could be `const`

File: nexus_producer/include/OptionalArgs.h [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  std::string filename;
+  std::string detSpecFilename;
+  std::string broker;
+  std::string instrumentName = "test";
+  std::string compression;
+  bool slow = false;
+  bool quietMode = false;
+  bool singleRun = false;
+  int32_t fakeEventsPerPulse = 0;
+  float pulseRateInHz = 10;
```
Comment (rerpha) [inline nexus_producer/include/OptionalArgs.h]: is this used anymore?

File: nexus_producer/src/main.cpp [hunk 5]
```diff
@@ -1,59 +1,75 @@
...
                  "Used as prefix for topic names");
-  App.add_option("-m,--compression", compression,
+  App.add_option("-m,--compression", settings.compression,
                  "Compression option for Kafka messages");
-  App.add_option("-e,--fake_events_per_pulse", fakeEventsPerPulse,
+  App.add_option("-e,--fake_events_per_pulse", settings.fakeEventsPerPulse,
                  "Generates this number of fake events per pulse instead of "
                  "publishing real data from file");
-  App.add_flag("-s,--slow", slow,
+  App.add_option("-p,--pulse_rate_hz", settings.pulseRateInHz,
```
Comment (rerpha) [inline nexus_producer/src/main.cpp]: can this be removed?

**Pull request merged**